src/lib.rs

```diff
@@ -0,0 +1,450 @@
+extern crate sxd_document;
+
+use std::str;
+use std::string::String;
+use std::collections::HashMap;
+
+use sxd_document::{
+    parser,
+    Package,
+};
+use sxd_document::dom::{
+    Document,
+    Root,
+    ChildOfRoot,
+    Element,
+    ChildOfElement,
+};
+
+#[derive(Debug)]
+pub struct Program {
+    pub groups: Vec<StatementBody>
+}
+
+#[derive(PartialEq, Debug)]
+pub struct StatementBody {
+    pub blocks: Vec<Block>
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Block {
+    pub block_type: String,
+    pub id: String,
+    pub fields: Option<HashMap<String, FieldValue>>,
+    pub statements: Option<HashMap<String, StatementBody>>,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum FieldValue {
+    SimpleField(String),
+    ExpressionField(Block),
+}
+
+impl Program {
+    pub fn new() -> Self {
+        Self {
+            groups: Vec::new()
+        }
+    }
+}
+
+impl StatementBody {
+    pub fn new() -> Self {
+        Self {
+            blocks: Vec::new()
+        }
+    }
+}
+
+impl Block {
+    pub fn new() -> Self {
+        Self {
+            block_type: "".to_string(),
+            id: "".to_string(),
+            fields: None,
+            statements: None
+        }
+    }
+
+    pub fn add_field(&mut self, name: String, value: FieldValue) {
+        match self.fields {
+            Some(ref mut f) => {
+                f.insert(name.to_owned(), value);
+            },
+            None => {
+                let mut new_map: HashMap<String, FieldValue> = HashMap::new();
+                new_map.insert(name.to_owned(), value);
+                self.fields = Some(new_map);
+            }
+        }
+    }
+
+    pub fn get_field(&self, name: &str) -> Option<&FieldValue> {
+        match self.fields {
+            Some(ref fields) => {
+                fields.get(name)
+            },
+            None => None
+        }
+    }
+
+    pub fn add_statement_body(&mut self, name: String, body: StatementBody) {
+        match self.statements {
+            Some(ref mut s) => {
+                s.insert(name.to_owned(), body);
+            },
+            None => {
+                let mut new_map: HashMap<String, StatementBody> = HashMap::new();
+                new_map.insert(name.to_owned(), body);
+                self.statements = Some(new_map);
+            }
+        };
+    }
+
+    pub fn get_statement(&self, name: &str) -> Option<&StatementBody> {
+        match self.statements {
+            Some(ref statements) => {
+                statements.get(name)
+            },
+            None => None
+        }
+    }
+}
+
+
+// Utilities for creating Blockly data structures
+
+pub fn program_from_xml(xml: &str) -> Program {
```

Comment:
Just calling attention to the fact that `xml` is of type `&str`, which it should be. This function does not need to take ownership of the actual xml in memory. This is likely owned by something else in your program.

If you are reading xml in from a file, then the data is likely owned by some buffer that you passed into a file reading function.

If you are reading xml in from a string literal (like in your tests below) then the data is owned by the binary itself (and are loaded into memory when the actual binary is; since this data lives for the duration of the entire binary, the reference to it has `'static` lifetime).